File: recipes/claude/21-provision-sandbox-computer-use/python/computer_use_orchestrator.py
"""
Claude Computer Use Virtual Desktop Sandbox Orchestrator.
Demonstrates:
- Computer tool registration (`computer_20250124`).
- Screen coordinate handling and action dispatch (`mouse_move`, `left_click`, `type`, `key`, `screenshot`).
- Human-in-the-loop validation boundary for sensitive actions.
- Multi-turn execution loop feeding base64 screenshot results back to Claude.
"""

import os
import sys
import base64
import json
import logging
from typing import Dict, Any, List, Optional
from anthropic import Anthropic

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")
logger = logging.getLogger("computer-use-sandbox")

# 1. Mock Screen Interface (Simulates an Xvfb / VNC virtual desktop framebuffer)
class VirtualDesktop:
    def __init__(self, width: int = 1024, height: int = 768, display_num: int = 1):
        self.width = width
        self.height = height
        self.display_num = display_num
        self.cursor_x = 0
        self.cursor_y = 0
        self.action_history: List[Dict[str, Any]] = []

    def capture_screenshot_base64(self) -> str:
        """
        Returns a mock 1x1 transparent PNG encoded in base64.
        In production, replace with:
            subprocess.run(["scrot", "-z", "/tmp/screen.png"])
            with open("/tmp/screen.png", "rb") as f:
                return base64.b64encode(f.read()).decode("utf-8")
        """
        # Minimal valid 1x1 transparent PNG
        png_bytes = (
            b"\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x01\x00\x00\x00\x01"
            b"\x08\x06\x00\x00\x00\x1f\x15c4\x00\x00\x00\nIDATx\x9cc\x00\x01\x00\x00\x05"
            b"\x00\x01\r\n-\xb4\x00\x00\x00\x00IEND\xaeB`\x82"
        )
        return base64.b64encode(png_bytes).decode("utf-8")

    def execute_action(self, action: str, **kwargs) -> Dict[str, Any]:
        """Dispatches an action to the virtual desktop."""
        self.action_history.append({"action": action, "kwargs": kwargs})

        if action == "screenshot":
            logger.info("VirtualDesktop: Capturing screenshot at %sx%s", self.width, self.height)
            return {"screenshot_base64": self.capture_screenshot_base64()}

        elif action == "mouse_move":
            x, y = kwargs.get("coordinate", [0, 0])
            self.cursor_x = max(0, min(self.width, x))
            self.cursor_y = max(0, min(self.height, y))
            logger.info("VirtualDesktop: Mouse moved to (%s, %s)", self.cursor_x, self.cursor_y)
            return {"status": "success", "cursor": [self.cursor_x, self.cursor_y]}

        elif action in ("left_click", "right_click", "double_click", "triple_click", "middle_click"):
            logger.info("VirtualDesktop: Executing %s at current position (%s, %s)", action, self.cursor_x, self.cursor_y)
            return {"status": "success", "action": action}

        elif action == "type":
            text = kwargs.get("text", "")
            logger.info("VirtualDesktop: Typing text (length %d)", len(text))
            return {"status": "success", "characters_typed": len(text)}

        elif action == "key":
            text = kwargs.get("text", "")
            logger.info("VirtualDesktop: Pressing key combination '%s'", text)
            return {"status": "success", "key": text}

        elif action == "cursor_position":
            return {"coordinate": [self.cursor_x, self.cursor_y]}

        else:
            logger.warning("VirtualDesktop: Unsupported action '%s'", action)
            return {"status": "unsupported_action", "action": action}


# 2. Security Boundary and Human-In-The-Loop Gate
class SandboxSecurityGateway:
    SENSITIVE_ACTIONS = {"submit_order", "delete_file", "sudo", "transfer_funds"}

    @classmethod
    def evaluate_safety(cls, action: str, parameters: Dict[str, Any]) -> bool:
        """
        Evaluates whether an action requires human operator confirmation.
        Returns True if safe to proceed automatically, False if gate is tripped.
        """
        # Check text keystrokes for destructive shell commands
        text = parameters.get("text", "").lower()
        if "rm -rf" in text or "sudo" in text or ":(){ :|:& };:" in text:
            logger.warning("SECURITY ALERT: Destructive command detected in keystroke buffer: %s", text)
            return False
        return True


# 3. Agent Execution Loop
def run_computer_use_agent(goal: str, max_turns: int = 5):
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        logger.error("ANTHROPIC_API_KEY environment variable is required.")
        sys.exit(1)

    client = Anthropic(api_key=api_key)
    desktop = VirtualDesktop(width=1024, height=768, display_num=1)
    security_gateway = SandboxSecurityGateway()

    # Tool definition for computer use beta
    computer_tool = {
        "type": "computer_20250124",
        "name": "computer",
        "display_width_px": desktop.width,
        "display_height_px": desktop.height,
        "display_number": desktop.display_num,
    }

    messages: List[Dict[str, Any]] = [
        {"role": "user", "content": goal}
    ]

    turn = 0
    while turn < max_turns:
        turn += 1
        logger.info("Turn %d: Calling Anthropic API...", turn)

        response = client.messages.create(
            model="claude-3-7-sonnet-20250219",
            max_tokens=2048,
            tools=[computer_tool],
            betas=["computer-use-2025-01-24"],
            messages=messages,
        )

        logger.info("Response stop reason: %s", response.stop_reason)

        # Append assistant message to conversation history
        messages.append({"role": "assistant", "content": response.content})

        if response.stop_reason != "tool_use":
            logger.info("Agent finished execution with reason: %s", response.stop_reason)
            for block in response.content:
                if getattr(block, "type", "") == "text":
                    print(f"\nFinal Response:\n{block.text}")
            break

        # Process tool calls
        tool_results: List[Dict[str, Any]] = []
        for block in response.content:
            if getattr(block, "type", "") != "tool_use":
                continue

            tool_use_id = block.id
            tool_name = block.name
            tool_input = block.input

            logger.info("Tool invocation requested: %s (id: %s)", tool_name, tool_use_id)

            if tool_name == "computer":
                action = tool_input.get("action")
                # Human-in-the-loop safety gate
                if not security_gateway.evaluate_safety(action, tool_input):
                    logger.error("Action blocked by Security Gateway: %s", action)
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": tool_use_id,
                        "is_error": True,
                        "content": "SecurityPolicyViolation: Action rejected by human-in-the-loop safety gate."
                    })
                    continue

                # Execute action on virtual desktop
                exec_result = desktop.execute_action(action, **tool_input)

                if action == "screenshot":
                    # For screenshot, return image block inside tool_result
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": tool_use_id,
                        "content": [
                            {
                                "type": "image",
                                "source": {
                                    "type": "base64",
                                    "media_type": "image/png",
                                    "data": exec_result["screenshot_base64"]
                                }
                            }
                        ]
                    })
                else:
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": tool_use_id,
                        "content": json.dumps(exec_result)
                    })
            else:
                tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": tool_use_id,
                    "is_error": True,
                    "content": f"Unknown tool: {tool_name}"
                })

        # Append tool results as user turn
        messages.append({"role": "user", "content": tool_results})


if __name__ == "__main__":
    test_goal = os.environ.get("TEST_GOAL", "Open the terminal and check system date.")
    logger.info("Starting Computer Use Sandbox Orchestrator with goal: '%s'", test_goal)
    run_computer_use_agent(test_goal)

File: recipes/claude/21-provision-sandbox-computer-use/typescript/computer_use_orchestrator.ts
import Anthropic from "@anthropic-ai/sdk";

// Minimal valid 1x1 transparent PNG base64
const MOCK_SCREENSHOT_BASE64 =
  "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==";

interface DesktopActionInput {
  action: string;
  coordinate?: [number, number];
  text?: string;
  [key: string]: unknown;
}

class VirtualDesktop {
  public width: number;
  public height: number;
  public displayNumber: number;
  public cursorX: number = 0;
  public cursorY: number = 0;

  constructor(width = 1024, height = 768, displayNumber = 1) {
    this.width = width;
    this.height = height;
    this.displayNumber = displayNumber;
  }

  public executeAction(input: DesktopActionInput): Record<string, unknown> {
    const { action, coordinate, text } = input;

    if (action === "screenshot") {
      return { screenshot_base64: MOCK_SCREENSHOT_BASE64 };
    } else if (action === "mouse_move" && coordinate) {
      this.cursorX = Math.max(0, Math.min(this.width, coordinate[0]));
      this.cursorY = Math.max(0, Math.min(this.height, coordinate[1]));
      return { status: "success", cursor: [this.cursorX, this.cursorY] };
    } else if (["left_click", "right_click", "double_click"].includes(action)) {
      return { status: "success", action, cursor: [this.cursorX, this.cursorY] };
    } else if (action === "type" && text) {
      return { status: "success", characters_typed: text.length };
    } else if (action === "key" && text) {
      return { status: "success", key_pressed: text };
    }
    return { status: "unsupported_action", action };
  }
}

class SandboxSecurityGateway {
  public static evaluateSafety(action: string, input: DesktopActionInput): boolean {
    const text = (input.text || "").toLowerCase();
    if (text.includes("rm -rf") || text.includes("sudo") || text.includes(":(){ :|:& };:")) {
      return false;
    }
    return true;
  }
}

async function runComputerUseAgent(goal: string, maxTurns = 5) {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    console.error("Error: ANTHROPIC_API_KEY is required.");
    process.exit(1);
  }

  const client = new Anthropic({ apiKey });
  const desktop = new VirtualDesktop(1024, 768, 1);

  const computerTool: Anthropic.Beta.Messages.BetaToolComputerUse20250124 = {
    type: "computer_20250124",
    name: "computer",
    display_width_px: desktop.width,
    display_height_px: desktop.height,
    display_number: desktop.displayNumber,
  };

  const messages: Anthropic.Beta.Messages.BetaMessageParam[] = [
    { role: "user", content: goal },
  ];

  let turn = 0;
  while (turn < maxTurns) {
    turn++;
    console.log(`[Turn ${turn}] Invoking Anthropic API...`);

    const response = await client.beta.messages.create({
      model: "claude-3-7-sonnet-20250219",
      max_tokens: 2048,
      tools: [computerTool],
      betas: ["computer-use-2025-01-24"],
      messages,
    });

    console.log(`[Turn ${turn}] Stop reason: ${response.stop_reason}`);
    messages.push({ role: "assistant", content: response.content });

    if (response.stop_reason !== "tool_use") {
      console.log("Agent run completed.");
      break;
    }

    const toolResults: Anthropic.Beta.Messages.BetaToolResultBlockParam[] = [];

    for (const block of response.content) {
      if (block.type === "tool_use" && block.name === "computer") {
        const input = block.input as DesktopActionInput;
        const action = input.action;

        if (!SandboxSecurityGateway.evaluateSafety(action, input)) {
          console.warn(`[Security Alert] Action blocked by security gateway: ${action}`);
          toolResults.push({
            type: "tool_result",
            tool_use_id: block.id,
            is_error: true,
            content: "SecurityPolicyViolation: Action rejected by sandbox safety policy.",
          });
          continue;
        }

        const result = desktop.executeAction(input);

        if (action === "screenshot") {
          toolResults.push({
            type: "tool_result",
            tool_use_id: block.id,
            content: [
              {
                type: "image",
                source: {
                  type: "base64",
                  media_type: "image/png",
                  data: result.screenshot_base64 as string,
                },
              },
            ],
          });
        } else {
          toolResults.push({
            type: "tool_result",
            tool_use_id: block.id,
            content: JSON.stringify(result),
          });
        }
      }
    }

    messages.push({ role: "user", content: toolResults });
  }
}

const targetGoal = process.env.TEST_GOAL || "Launch browser and navigate to example.com";
runComputerUseAgent(targetGoal).catch(console.error);
